pkg/bots/tgbot/utils.go

```diff
@@ -82,11 +86,51 @@ func FullNameFromFirstAndLastName(firstName, lastName string) string {
 }
 
 func ExtractTextFromMessage(message *tgbotapi.Message) string {
-	if message.Caption != "" {
-		return message.Caption
+	text := lo.Ternary(message.Caption != "", message.Caption, message.Text)
+
+	type MarkdownLink struct {
+		Markdown []uint16
+		Start    int
+		End      int
+	}
+
+	textUtf16 := utf16.Encode([]rune(text))
+	links := lop.Map(message.Entities, func(entity tgbotapi.MessageEntity, i int) MarkdownLink {
+		startIndex := entity.Offset
+		endIndex := startIndex + entity.Length
+		var title string
+		var href string
+		if entity.Type == "url" {
+			href = string(utf16.Decode(textUtf16[startIndex:endIndex]))
+			result, err := LinkPreview.PreviewLink(href, nil)
+			if err != nil {
+				return MarkdownLink{[]uint16{}, -1, -1}
+			}
+			title = result.Title
+		} else if entity.Type == "text_link" {
+			title = string(utf16.Decode(textUtf16[startIndex:endIndex]))
+			href = entity.URL
+		} else {
+			return MarkdownLink{[]uint16{}, -1, -1}
+		}
+		unescaped, err := url.QueryUnescape(href)
+		if err == nil {
```

Comment:
What about err != nil?

Reply:
Unescaping the link is to make it interpretable for AI as well as save tokens. Note that the link is not necessarily a URI encoded one, and errors occur when `%` has no succeeding hex value. A link may contain many `%` without being encoded. And if it is unescapable, we just keep it. Or, is it a convention to write `if err != nil { } else { // do something }`?

Reply:
Nah, it is ok to omit the err.

But I have checked the source code of `url.QueryEscape(...)` and find out that `unescaped` is always empty if any err occurred, so you might have to deal with the empty string I guess?

Reply:
`if err != nil` then `href` won't be modified and href is kept as it is.

Reply:
Understood.